### PR TITLE
Advanced Mapping : Broken Field References in Advanced Mapping Fix

### PR DESCRIPTION
**Pull request by voduyemi**
Added a mechanism to manage invalid field mappings in the Advanced Mapping Page in NPSP Settings

WI: [W-7780916](https://gus.lightning.force.com/lightning/r/a07B0000008EI2OIAW/view)
# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata
Custom Labels:
bdiOMUIFieldMappingProblemHeader
bdiOMUIFieldMappingProblemMessagePart1
bdiOMUIFieldMappingProblemMessagePart2
# Deleted Metadata

### CODE REVIEW

File: src/classes/BDI_MappingServiceAdvanced.cls [hunk 1]
```diff
@@ -317,6 +235,108 @@ public with sharing class BDI_MappingServiceAdvanced implements BDI_MappingServi
         return fieldMappingInstance.fieldMappingSet;
     }
 
+    private List<Data_Import_Object_Mapping__mdt> getDataImportObjectMappingMetadata (
+        Data_Import_Field_Mapping_Set__mdt diFieldMappingSet) {
+        return [
```
Comment (psadie-sfdo) [inline src/classes/BDI_MappingServiceAdvanced.cls:240]: Need to align the formatting for these queries as per the apex style guide https://salesforce.quip.com/TkjIACNQIURb#FGVACAY5bs9

File: src/lwc/bdiFieldMappingModal/bdiFieldMappingModal.js [hunk 2]
```diff
@@ -215,7 +215,7 @@ export default class bdiFieldMappingModal extends LightningElement {
                 this.modalMode = 'new';
                 this.fieldMapping = {};
             }
-
+            console.log(JSON.stringify(this.fieldMapping));
```
Comment (psadie-sfdo) [inline src/lwc/bdiFieldMappingModal/bdiFieldMappingModal.js:218]: Log heads up

File: src/lwc/bdiObjectMappings/bdiObjectMappings.js [hunk 3]
```diff
@@ -364,28 +355,6 @@ export default class bdiObjectMappings extends LightningElement {
         }
     }
 
-    /*******************************************************************************
```
Comment (psadie-sfdo) [inline src/lwc/bdiObjectMappings/bdiObjectMappings.js:367]: 👍

File: src/classes/BDI_ManageAdvancedMappingCtrl.cls [hunk 4]
```diff
@@ -202,27 +216,42 @@ public class BDI_ManageAdvancedMappingCtrl {
...
+                    fieldMappingWrappers.add(wrapper);
                 }
             }
         }
-
         return fieldMappingWrappers;
     }
 
+    /**
+     * @description
```
Comment (psadie-sfdo) [inline src/classes/BDI_ManageAdvancedMappingCtrl.cls:235]: Missing description

File: src/classes/BDI_ManageAdvancedMappingCtrl.cls [hunk 5]
```diff
@@ -202,27 +216,42 @@ public class BDI_ManageAdvancedMappingCtrl {
...
+     * @param bdiMsAdv
+     * @param objectName
+     *
+     * @return {BDI_FieldMapping[]} fieldMappings
+     */
+    private static BDI_FieldMapping[] injectBrokenFieldMappings (
+        BDI_MappingServiceAdvanced bdiMsAdv, String objectName) {
```
Comment (psadie-sfdo) [inline src/classes/BDI_ManageAdvancedMappingCtrl.cls:242]: Do we need to pass `bdiMsAdv` into this method?

File: src/classes/BDI_MappingServiceAdvanced.cls [hunk 6]
```diff
@@ -317,6 +235,108 @@ public with sharing class BDI_MappingServiceAdvanced implements BDI_MappingServi
...
+                    BDI_FieldMapping fieldMapping;
+                    try {
+                        fieldMapping = new BDI_FieldMapping(difm);
+                    } catch (UTIL_Describe.SchemaDescribeException sde) {
+                        if (sde.getMessage().contains('Invalid')) {
+                            constructBrokenFieldMappings(difm);
+                            continue;
+                        }
+                    }
+                    BDI_FieldMapping[] tempFieldMappings = new BDI_FieldMapping[]{};
```
Comment (psadie-sfdo) [inline src/classes/BDI_MappingServiceAdvanced.cls:302]: Would it make sense to pull lines 302 - 314 into its own named method?

File: src/classes/BDI_MappingServiceAdvanced.cls [hunk 7]
```diff
@@ -415,4 +437,31 @@ public with sharing class BDI_MappingServiceAdvanced implements BDI_MappingServi
 
         return fieldMapping;
     }
+
+    private void constructBrokenFieldMappings (Data_Import_Field_Mapping__mdt difm) {
```
Comment (psadie-sfdo) [inline src/classes/BDI_MappingServiceAdvanced.cls:441]: Any particular reason we didn't want to perform this type of logic inside the `BDI_FieldMapping` constructor?

File: src/classes/BDI_ObjectMapping.cls [hunk 8]
```diff
@@ -84,11 +85,20 @@ public virtual with sharing class BDI_ObjectMapping {
...
+            this.Imported_Record_Field_Name = UTIL_Describe.getFieldName(SObjectType.DataImport__c.Name,
+                objectMapping.Imported_Record_Field_Name__c);
+            this.Imported_Record_Status_Field_Name = UTIL_Describe.getFieldName(SObjectType.DataImport__c.Name,
+                objectMapping.Imported_Record_Status_Field_Name__c);
+            this.Object_API_Name = UTIL_Describe.getObjectName(objectMapping.Object_API_Name__c);
+        } catch (UTIL_Describe.SchemaDescribeException sde) {
+            if (sde.getMessage().contains('Invalid')) {
+                this.Is_Broken = true;
+            }
+        } finally {
```
Comment (psadie-sfdo) [inline src/classes/BDI_ObjectMapping.cls]: Do we need this empty `finally`?

File: src/lwc/bdiObjectMappings/bdiObjectMappings.html [hunk 9]
```diff
@@ -33,6 +32,26 @@ <h1 class="slds-text-heading_medium slds-text-heading_medium_extension slds-m-bo
                 <template if:true={shouldRender}>
                     <h2 class="slds-text-heading_small slds-text-heading_small_extension slds-m-bottom_medium">{customLabels.bdiOMUIObjectGroupsTitle}</h2>
 
+                    <template if:true={hasBrokenMetadataReferences}>
+                        <div class="slds-size_1-of-1">
+                            <c-util-page-level-message title='Check Field Mappings'
```
Comment (psadie-sfdo) [inline src/lwc/bdiObjectMappings/bdiObjectMappings.html:37]: Are we adding custom labels in a separate work items for the title and subtitle here?

File: src/lwc/utilCommon/utilCommon.js [hunk 10]
```diff
@@ -528,6 +528,15 @@ const stripNamespace = (apiName , namespacePrefix) => {
     return apiNameParts[1];
 }
 
+/*******************************************************************************
+ * @description Parse proxy objects for debugging, mutating, etc
+ *
+ * @param {object} obj: Object to be parsed
+ */
+const parse = (obj) => {
```
Comment (psadie-sfdo) [inline src/lwc/utilCommon/utilCommon.js:536]: We have an existing method called `mutable` that does the same thing here. Could we align to using one or the other? I have no preference regarding the name fwiw

**Review by psadie-sfdo — COMMENTED** [review]
A few more comments. Additionally what do you think about bringing up the api versions of touched files to 49?

File: src/lwc/bdiObjectMappings/bdiObjectMappings.js [hunk 11]
```diff
@@ -176,20 +152,25 @@ export default class bdiObjectMappings extends LightningElement {
...
+     * @description Extracts broken field mapping references from retrieved
+     * object mappings
+     * @param objectMappings
+     */
+    processBrokenMappingReferences (objectMappings) {
+        this.brokenMappings = [];
+        objectMappings.forEach(mapping => {
+            if (mapping.hasOwnProperty('Field_Mappings')) {
+                mapping.Field_Mappings.forEach(fieldMapping => {
+                    if (fieldMapping.Is_Broken && !fieldMapping.Is_Deleted) {
```
Comment (ptewson-sfdo) [inline src/lwc/bdiObjectMappings/bdiObjectMappings.js:165]: Definitely see some overlap here with the other processBrokenFieldMappingReferences function but maybe we can't avoid it ...

**Review by ptewson-sfdo — COMMENTED** [review]
I'm excited about this! Left several comments, some more questions than comments. Let me know if anything doesn't make sense!

File: src/classes/BDI_ManageAdvancedMappingCtrl.cls [hunk 12]
```diff
@@ -119,23 +119,37 @@ public class BDI_ManageAdvancedMappingCtrl {
...
             if(predecessorObjMapping != null) {
                 objMapping.Predecessor_Label_Name = predecessorObjMapping.MasterLabel;
             }
         }
-
         return objMappings;
     }
 
+    @AuraEnabled
+    public static AdvancedMappingObjectData getAdvancedMappingObjectData () {
```
[the remark this answers is not in the document]
Comment (voduyemi) [reply src/classes/BDI_ManageAdvancedMappingCtrl.cls:152]: Done!

File: src/classes/BDI_ManageAdvancedMappingCtrl.cls [hunk 13]
```diff
@@ -119,23 +119,37 @@ public class BDI_ManageAdvancedMappingCtrl {
...
+        return new AdvancedMappingObjectData(getObjectMappings(), getObjectOptions());
+    }
+
+    @AuraEnabled(Cacheable = true)
+    public static AdvancedMappingFieldData getAdvancedMappingFieldsData (String targetObjectApiName,
```
[the remark this answers is not in the document]
Comment (voduyemi) [reply src/classes/BDI_ManageAdvancedMappingCtrl.cls:164]: Done.

File: src/classes/BDI_ManageAdvancedMappingCtrl.cls [hunk 14]
```diff
@@ -802,4 +829,33 @@ public class BDI_ManageAdvancedMappingCtrl {
             this.npspNamespace = UTIL_Namespace.HARDCODED_NPSP_NAMESPACE;
         }
     }
+
+    public class AdvancedMappingObjectData {
```
[the remark this answers is not in the document]
Comment (voduyemi) [reply src/classes/BDI_ManageAdvancedMappingCtrl.cls:854]: Done!

File: src/classes/BDI_ManageAdvancedMappingCtrl.cls [hunk 15]
```diff
@@ -802,4 +829,33 @@ public class BDI_ManageAdvancedMappingCtrl {
...
+        public List<ComboboxOption> objectOptions;
+
+        public AdvancedMappingObjectData (BDI_ObjectMapping [] objectMappings,
+            List<ComboboxOption> objectOptions) {
+            this.objectMappings = objectMappings;
+            this.objectOptions = objectOptions;
+        }
+    }
+
+    public class AdvancedMappingFieldData {
```
[the remark this answers is not in the document]
Comment (voduyemi) [reply src/classes/BDI_ManageAdvancedMappingCtrl.cls:871]: Done!

File: src/classes/BDI_MappingServiceAdvanced.cls [hunk 16]
```diff
@@ -317,6 +235,108 @@ public with sharing class BDI_MappingServiceAdvanced implements BDI_MappingServi
...
+            WHERE Is_Deleted__c = false
+                AND Target_Object_Mapping__r.Is_Deleted__c = false)
+            FROM Data_Import_Object_Mapping__mdt
+            WHERE Data_Import_Object_Mapping_Set__c =: diFieldMappingSet.Data_Import_Object_Mapping_Set__c
+                AND Is_Deleted__c = false
+            ORDER BY MasterLabel ASC
+        ];
+    }
+
+    private List<Data_Import_Field_Mapping_Set__mdt> getDataImportFieldMappingSets () {
```
[the remark this answers is not in the document]
Comment (voduyemi) [reply src/classes/BDI_MappingServiceAdvanced.cls:282]: Done!

**Review by ptewson-sfdo — APPROVED** [review]
Nice work, Victor! This looks good!

**Comment by melissabarber** [comment]
@voduyemi, @psadie-sfdo, & @ptewson-sfdo I reviewed the custom labels and this PR is good to go from a docs point of view. Thanks!

**Pull request merged**